bokehjs/src/lib/core/util/templating.ts

```diff
@@ -36,8 +39,28 @@ export function replace_placeholders(str: string, data_source: ColumnarDataSourc
     let value: any
     if (name[0] == "$")
       value = special_vars[name.substring(1)]
+
+    const column = data_source.get_column(name)
+
+    if (!isNumber(i)) { // An ImageIndex
+      if (column != null) {
+        const data = column[i['index']]
+        if (isTypedArray(data) || isArray(data)) {
+          if (isArray(data[0])) { // Array of arrays format
+            let row : any
+            row = data[i['dim2']]
+            value = row[i['dim1']]
+          }
+          else {
+            value = data[i['flat_index']]
+          }
+        }
+        else {
+          value = data
+        }
+      }
+    }
```

Comment:
This function is now pushing 100 lines, it would be good to split it up in to some sub-functions that are also amendable to unit tests

Reply:
Additionally, and in particular, the code to apply a formatter should apply to images as well, so that could be pulled out and put under test, then each image/non-image path could also probably be its own helper function